src/platform/windows/mod.rs

```diff
@@ -0,0 +1,1482 @@
+                kernel32::CancelIoEx(self.handle, self.ov_ptr());
+                self.read_in_progress = false;
+            }
+        }
+    }
+
+    fn ensure_is_reading(&mut self) -> Result<(),WinError> {
+        self.start_read()
+    }
+
+    // Called when we receive an IO Completion Packet for this handle.
+    fn notify_completion(&mut self, err: u32) -> Result<(),WinError> {
+        dd2!("[$ {:?}:{:?}] notify_completion", self.iocp, self.handle);
+
+        if err == winapi::ERROR_BROKEN_PIPE {
+            assert!(!self.closed, "we shouldn't get an async BROKEN_PIPE after we already got one");
+            self.closed = true;
+            return Ok(());
+        }
+
+        let nbytes = self.ov.InternalHigh as u32;
+        let offset = self.ov.Offset;
+
+        assert!(offset == 0);
+
+        // if the remote end closed...
+        if err != winapi::ERROR_SUCCESS {
+            // This should never happen
+            panic!("[$ {:?}:{:?}] *** notify_completion: unhandled error reported! {}", self.iocp, self.handle, err);
+        }
+
+        unsafe {
+            let new_size = self.read_buf.len() + nbytes as usize;
+            dd!("nbytes: {}, offset {}, buf len {}->{}, capacity {}",
+                nbytes, offset, self.read_buf.len(), new_size, self.read_buf.capacity());
+            assert!(new_size <= self.read_buf.capacity());
+            self.read_buf.set_len(new_size);
+        }
+
+        Ok(())
+    }
+
+    // kick off an asynchronous read
+    fn start_read(&mut self) -> Result<(),WinError> {
+        if self.read_in_progress || self.closed {
+            return Ok(());
+        }
+
+        dd2!("[$ {:?}:{:?}] start_read ov {:?}", self.iocp, self.handle, self.ov_ptr());
+
+        let buf_len = self.read_buf.len();
+        let mut buf_cap = self.read_buf.capacity();
+        let mut bytes_read: u32 = 0;
+
+        // if the buffer is full, add more capacity
+        if buf_cap == buf_len {
+            self.read_buf.reserve(match buf_cap {
+                0 => READ_BUFFER_SIZE,
+                1...READ_BUFFER_MAX_GROWTH => buf_cap,
+                _ => READ_BUFFER_MAX_GROWTH
+            });
+            buf_cap = self.read_buf.capacity();
+        }
+
+        // issue the read to the buffer, at the current length offset
+        unsafe {
+            self.reset_overlapped();
+            let buf_ptr = self.read_buf.as_mut_ptr() as LPVOID;
+            let max_read_bytes = buf_cap - buf_len;
+            let ok = kernel32::ReadFile(self.handle,
+                                        buf_ptr.offset(buf_len as isize),
+                                        max_read_bytes as u32,
+                                        &mut bytes_read,
+                                        self.ov_ptr());
+
+            // ReadFile can return TRUE; if it does, an IO completion
+            // packet is still posted to any port, and the OVERLAPPED
+            // structure has the IO operation flagged as complete.
```

Comment:
Normally, for an async operation, a call like `ReadFile` would return `FALSE`, and the error code would be `ERROR_IO_PENDING`.  But in some situations, `ReadFile` can complete synchronously (returns `TRUE`).  Even if it does, a notification that the IO completed is still sent to the IO completion port that this handle is part of, meaning that we don't have to do any special handling for sync-completed operations.  (I just added that as a comment to the file.)